File: Questions/Java/Question3.java
public class Question3 {
    public class BinarySearch {
        static int binarySearch(int[] arr, int target) {

            int start = 0;
            int end = arr.length - 1;

            while (start <= end) {
                // write
                // your
                // code
                // here
            }

            return -1;
        }
    }

    public static void main(String[] args) {
        int[] array1 = { 1, 2, 3, 4, 5, 6, 7, 8, 9, 10 };
        int target1 = 5;
        System.out.println("Test Case 1: " + BinarySearch.binarySearch(array1, target1)); // Expected output: 4

        // Test Case 2: Element found at the beginning
        int[] array2 = { 1, 2, 3, 4, 5, 6, 7, 8, 9, 10 };
        int target2 = 1;
        System.out.println("Test Case 2: " + BinarySearch.binarySearch(array2, target2)); // Expected output: 0

        // Test Case 3: Element found at the end
        int[] array3 = { 1, 2, 3, 4, 5, 6, 7, 8, 9, 10 };
        int target3 = 10;
        System.out.println("Test Case 3: " + BinarySearch.binarySearch(array3, target3)); // Expected output: 9

        // Test Case 4: Element not found
        int[] array4 = { 1, 2, 3, 4, 5, 6, 7, 8, 9, 10 };
        int target4 = 15;
        System.out.println("Test Case 4: " + BinarySearch.binarySearch(array4, target4)); // Expected output: -1

        // Test Case 5: Empty array
        int[] array5 = {};
        int target5 = 5;
        System.out.println("Test Case 5: " + BinarySearch.binarySearch(array5, target5)); // Expected output: -1
    }
}

File: Questions/C/Question3.c
// Question 3
// Program to Search array in O(logn) complexity.

#include <stdio.h>

// Function to perform binary search on a sorted array
int search(int arr[], int size, int target)
{
    int start = 0;
    int end = size - 1;

    while (start <= end)
    {
        // Write
        // Your
        // Code
        // Here
    }

    return -1;
}

int main()
{
    struct Question3 q3;

    int array1[] = {1, 2, 3, 4, 5, 6, 7, 8, 9, 10};

    // Test Case 1
    int target1 = 5;
    printf("Test Case 1: %d\n", search(array1, sizeof(array1) / sizeof(array1[0]), target1));

    // Test Case 2
    int array2[] = {1, 2, 3, 4, 5, 6, 7, 8, 9, 10};
    int target2 = 1;
    printf("Test Case 2: %d\n", search(array2, sizeof(array2) / sizeof(array2[0]), target2));

    // Test Case 3
    int array3[] = {1, 2, 3, 4, 5, 6, 7, 8, 9, 10};
    int target3 = 10;
    printf("Test Case 3: %d\n", search(array3, sizeof(array3) / sizeof(array3[0]), target3));

    // Test Case 4
    int array4[] = {1, 2, 3, 4, 5, 6, 7, 8, 9, 10};
    int target4 = 15;
    printf("Test Case 4: %d\n", search(array4, sizeof(array4) / sizeof(array4[0]), target4));

    // Test Case 5
    int array5[] = {};
    int target5 = 5;
    printf("Test Case 5: %d\n", search(array5, sizeof(array5) / sizeof(array5[0]), target5));

    return 0;
}
